STM32L4: EXTI, CRC, SPI, DMA, IWDG, RTC, CRS, HSI48, USB

br101 opened the pull request:
Adding more support for L4, some parts taken from https://github.com/libopencm3/libopencm3/pull/854
All tested on STM32L433CC.

karlp commented:
I think there's another pr for maaking a CRC-v2 driver, so I'll skip that for now.  most of the rest look good.

karlp commented on include/libopencm3/stm32/l4/rtc.h:
```diff
@@ -0,0 +1,36 @@
+/** @defgroup rtc_defines RTC Defines
+
+@brief <b>Defined Constants and Types for the STM32L1xx RTC</b>
```
typo

karlp commented on include/libopencm3/stm32/l4/rcc.h:
```diff
@@ -133,6 +134,10 @@ Twelve frequency ranges are available: 100 kHz, 200 kHz, 400 kHz, 800 kHz,
+/* --- RCC_CRRCR values ---------------------------------------------------- */
+
+#define RCC_CRRCR_HSI48ON             (1 << 0)
+#define RCC_CRRCR_HSI48RDY            (1 << 1)
```
thhe _commit_ message for this one is wrong. referring to importing and testing is for the PR, not the commit message, othherwise looks good

karlp commented on include/libopencm3/stm32/l4/rcc.h:
```diff
@@ -587,7 +587,7 @@ Twelve frequency ranges are available: 100 kHz, 200 kHz, 400 kHz, 800 kHz,
 #define RCC_CCIPR_ADCSEL_SHIFT		28
 
-#define RCC_CCIPR_CLK48SEL_NONE		0
+#define RCC_CCIPR_CLK48SEL_HSI48	0
```
is this actually a problem with different L4s? looks suspect

br101 replied:
I checked it against both reference manuals I found for the L4 (RM0394 and RM0351)

karlp replied:
Ok, found it.  I wrote that from RM0351 rev1, this bit is listed has having been updated in revision 5, the current revision :) so yeah, looks good now :)

karlp commented on include/libopencm3/stm32/l4/exti.h:
```diff
@@ -0,0 +1,49 @@
+/** @defgroup exti_defines EXTI Defines
+ *
+ * @brief <b>Defined Constants and Types for the STM32F3xx External Interrupts
```
typo

karlp commented on include/libopencm3/stm32/l4/exti.h:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @version 1.0.0
+ *
+ * @date 28 December 2018
```
a little optimistic :)

karlp commented on include/libopencm3/stm32/l4/syscfg.h:
```diff
@@ -0,0 +1,115 @@
+ *
+ * @version 1.0.0
+ *
+ * @date 28 December 2018
```
bad date again

br101 commented:
Thanks for the review and sorry for the typos... I updated the pull request and fixed some more typos, also reviewed against pull request https://github.com/libopencm3/libopencm3/pull/617 and added more syscfg.

I still wonder about the EXTI stuff... all above 32 are now broken or would have to be done by manually writing the EXTI_XXX2 registers - should I do it as in https://github.com/libopencm3/libopencm3/pull/617/commits/1c647124e3093dc22011efd4b5fb854bbda0e20c (allthough that one seems broken too, but you'll get the idea)?

karlp commented:
Just leave this as is for a bit, it's a really nice well separated set of changes.  I'll probably cherry pick a bunch of them in pieces as I go, but this is a really nice looking branch thanks!

0xdec commented on include/libopencm3/stm32/l4/memorymap.h:
```diff
@@ -56,7 +56,10 @@
+#define CRS_BASE			(PERIPH_BASE_APB1 + 0x6000)
 #define CAN1_BASE			(PERIPH_BASE_APB1 + 0x6400)
+#define USB_DEV_FS_BASE			(PERIPH_BASE_APB1 + 0x6800)
+#define USB_PMA_BASE			(PERIPH_BASE_APB1 + 0x6c00)
```
Minor nitpick: fix tab spacing

br101 replied:
i think it is... checkout the raw file... am i missing something???

0xdec replied:
Oh you're right, for some reason it was showing up wrong in the diff. Sorry! Hope this gets merged soon!

br101 commented:
Hi! Is there anything I can do to get this merged?

karlp commented:
reminding me helpd.  I've not been working much on this, but I guess it's time to get back onto it again :)

0xdec commented:
@karlp are you the only active maintainer? I'd love to support the project in addition to PRs, is there some way to donate (patreon etc)?

karlp commented on include/libopencm3/stm32/l4/syscfg.h:
```diff
@@ -0,0 +1,122 @@
+#define SYSCFG_CFGR2_FPU_IE3		(1 << 29)
+#define SYSCFG_CFGR2_FPU_IE2		(1 << 28)
+#define SYSCFG_CFGR2_FPU_IE1		(1 << 27)
+#define SYSCFG_CFGR2_FPU_IE0		(1 << 26)
```
really not sure this is a useful way of defining them. (the fpu bits here I mean)

br101 replied:
i made a mistake: this is  SYSCFG_CFGR1_* anyhow. will update the PR.

As for the FPU bits, should I try to find short names according to
FPU_IE[5]: Inexact interrupt enable
FPU_IE[4]: Input denormal interrupt enable
FPU_IE[3]: Overflow interrupt enable
FPU_IE[2]: underflow interrupt enable
FPU_IE[1]: Divide-by-zero interrupt enable
FPU_IE[0]: Invalid operation interrupt enable
?

karlp replied:
Maybe even just leave them out for now.  Perhaps just list their size and that they exist? Sometimes it's best to get the file started, without trying to get everything in.

br101 replied:
Well, I find naming them like I did in my last update makes most sense:
```
#define SYSCFG_CFGR1_FPU_IE_INEXACT	(1 << 31)
#define SYSCFG_CFGR1_FPU_IE_DENORMAL	(1 << 30)
#define SYSCFG_CFGR1_FPU_IE_OVERFLOW	(1 << 29)
#define SYSCFG_CFGR1_FPU_IE_UNDERFLOW	(1 << 28)
#define SYSCFG_CFGR1_FPU_IE_DIVZERO	(1 << 27)
#define SYSCFG_CFGR1_FPU_IE_INVALID	(1 << 26)
```

Or... if you prefer, we could only do
```
/* FPU interrupt enable (x from 0 to 5)
#define SYSCFG_CFGR1_FPU_IE(x)		(1 << (26 + x))
```

In any case, for me either way is fine...

karlp commented on include/libopencm3/stm32/l4/syscfg.h:
```diff
@@ -0,0 +1,122 @@
+
+/* --- SYSCFG_EXTICR Values -------------------------------------------------*/
+
+#define SYSCFG_EXTICR_SKIP		4
```
Is this meant to be saying how big each section of the register is?  it's a very uncommon way of doing it

br101 replied:
F0 has the same define, but in both cases (F0, L4) it's not used and i think exti_select_source() takes care of the 4 bit alignment... so i'll just remove it

karlp commented:
The EXTI commit is a little "too" standalone.  It actualyl requires the syscfg to go in first.

karlp commented on lib/stm32/l4/rcc.c:
```diff
@@ -326,7 +326,7 @@ void rcc_set_hpre(uint32_t hpre)
 	uint32_t pllq, uint32_t pllr)
 {
-	RCC_PLLCFGR = (pllm << RCC_PLLCFGR_PLLM_SHIFT) |
+	RCC_PLLCFGR = (RCC_PLLCFGR_PLLM(pllm) << RCC_PLLCFGR_PLLM_SHIFT) |
```
I don't like this one, it makes it different in parameters you pass compared to f2/f4/f7

br101 replied:
are you referring to source?
for f2/f4/f7 you have two different functions `rcc_set_main_pll_hsi()` and `rcc_set_main_pll_hse()`
for l4 there is MSI, HSI16 and HSE so we would have to make three equivalent functions...

or to RCC_PLLCFGR_PLLM(pllm)? my reasoning for this is that when you look at the clock tree and you decide for a M value it's not clear that the register expects the value - 1.

karlp replied:
I'm talking about the RCC_PLLCFGR_PLLM() macro indeed.  I see now looking at the f4 reference manual again what you're doing, and yeah, this is fine.  This can just come in by itself I think.

karlp replied:
I see thsi commit is already standalone. I'll pull it into my staging.

karlp commented on include/libopencm3/stm32/l4/irq.json:
```diff
@@ -81,7 +81,8 @@
         "rng",
-        "fpu"
+        "fpu",
+        "crs"
```
if you're going to do updates for teh L496 and similar, why not add all of them? :)

br101 replied:
i am working with L433 and was just looking at RM0394 at this moment. can't we add the other parts later?

karlp commented:
I've pulled _most_ of this into my karlp-staging-2018.02 branch.  I've left out exti/syscfg, and all the later rcc helpers.  Only compile tested at this point

karlp commented:
@0xdec I'm the only presently active maintainer, apparently yes.  Patreon wouldn't make any difference for me personally, it's not about money, it's just time.  (and not in a, I don't have any time because I need to make money kinda way, I just don't prioritize this "high enough")

The best thing you can do is take commits that have floundered, solve the problems with them, clean them up.  Review incoming patches so I don't have to, Write test cases!  Write more test cases!  All those sorts of things are good and helpful.

br101 commented:
Updated the SYSCFG commit according to your comments

karlp commented on lib/stm32/l4/rcc.c:
```diff
@@ -369,4 +369,15 @@ void rcc_set_msi_range_standby(uint32_t msi_range)
 	RCC_CSR = reg;
 }
 
+void rcc_pll_output_enable(uint32_t pllout)
```
doxygen on what this parameter means?  Is it actually a boolean? or what?

karlp commented on lib/stm32/l4/rcc.c:
```diff
@@ -369,4 +369,15 @@ void rcc_set_msi_range_standby(uint32_t msi_range)
+	RCC_PLLCFGR |= pllout;
+}
+
+void rcc_set_clock48_source(uint32_t clksel)
```
same here please

karlp commented:
I've just pushed ~most of this, up to the last rcc helpers and leaving out the flash programming one for now.  Can you please take what you still have and rebase on top of master and make a new PR for it please?
